File: frontend/sfx.js
/* v1.5.1/v1.6 — Sound effects (Web Audio, synthesized: no files, works offline).
 * Sfx.play(name):
 *   welcome  — short instrumental jingle on app start (after login)
 *   success  — sale completed / activation        add     — item added to cart
 *   error    — soft low double-tone               login   — rising sweep
 *   alert    — gentle two-note chime              note    — notification / bottom alert
 *   void     — invoice voided (descending)        hold    — invoice parked
 *   resume   — invoice restored                   step / click / ready / install / exit
 * All levels are deliberately low (a shop till must not be loud); master volume
 * "sm.sfx.vol" (0..1, default 0.35); "sm.sfx"="0" disables. */
(function () {
  "use strict";
  let ctx = null, master = null;
  const enabled = () => localStorage.getItem("sm.sfx") !== "0";
  const vol = () => { const v = parseFloat(localStorage.getItem("sm.sfx.vol")); return Number.isFinite(v) ? Math.min(1, Math.max(0, v)) : 0.35; };
  function ac() {
    if (ctx) return ctx;
    const AC = window.AudioContext || window.webkitAudioContext; if (!AC) return null;
    ctx = new AC();
    master = ctx.createGain(); master.gain.value = vol();
    // soft low-pass so nothing sounds harsh
    const lp = ctx.createBiquadFilter(); lp.type = "lowpass"; lp.frequency.value = 3800; lp.Q.value = 0.4;
    master.connect(lp).connect(ctx.destination);
    return ctx;
  }
  // unlock on first user gesture (browser autoplay policy)
  ["pointerdown", "keydown"].forEach((ev) => window.addEventListener(ev, () => { const c = ac(); if (c && c.state === "suspended") c.resume().catch(() => {}); }, { passive: true }));

  function tone(c, { f = 440, t0 = 0, d = 0.15, type = "sine", g = 0.3, slide = null, a = 0.012 }) {
    const o = c.createOscillator(), env = c.createGain();
    o.type = type; o.frequency.setValueAtTime(f, c.currentTime + t0);
    if (slide) o.frequency.exponentialRampToValueAtTime(slide, c.currentTime + t0 + d);
    env.gain.setValueAtTime(0.0001, c.currentTime + t0);
    env.gain.exponentialRampToValueAtTime(g, c.currentTime + t0 + a);
    env.gain.exponentialRampToValueAtTime(0.0001, c.currentTime + t0 + d);
    o.connect(env).connect(master); o.start(c.currentTime + t0); o.stop(c.currentTime + t0 + d + 0.02);
  }
  // simple pluck used by the welcome jingle (triangle + quiet octave)
  const pluck = (c, f, t0, d = 0.5, g = 0.16) => { tone(c, { f, t0, d, type: "triangle", g, a: 0.006 }); tone(c, { f: f * 2, t0, d: d * 0.6, type: "sine", g: g * 0.35, a: 0.004 }); };
  const N = { C4: 261.63, D4: 293.66, E4: 329.63, F4: 349.23, G4: 392, A4: 440, B4: 493.88, C5: 523.25, D5: 587.33, E5: 659.25, F5: 698.46, G5: 783.99, A5: 880, B5: 987.77, C6: 1046.5 };
  const P = {
    // ~3.5 s instrumental jingle (I–V–vi–IV feel), quiet
    welcome: (c) => {
      const seq = [[N.C5, 0], [N.E5, .18], [N.G5, .36], [N.C6, .54], [N.B5, .9], [N.G5, 1.08], [N.A5, 1.35], [N.C6, 1.53], [N.E5, 1.8], [N.F5, 1.98], [N.A5, 2.16], [N.G5, 2.55]];
      seq.forEach(([f, t]) => pluck(c, f, t, 0.55, 0.14));
      [[N.C4, 0], [N.G4, .9], [N.A4, 1.35], [N.F4, 1.98], [N.C4, 2.55]].forEach(([f, t]) => tone(c, { f, t0: t, d: 0.9, type: "sine", g: 0.09, a: 0.03 }));
      pluck(c, N.C6, 2.95, 1.2, 0.12); tone(c, { f: N.E5, t0: 2.95, d: 1.2, type: "sine", g: 0.06, a: 0.05 });
    },
    // cheerful major arpeggio — sale completed / activation
    success: (c) => { [N.C5, N.E5, N.G5, N.C6].forEach((f, i) => tone(c, { f, t0: i * 0.09, d: 0.22, type: "triangle", g: 0.16 })); tone(c, { f: N.G5 * 2, t0: 0.36, d: 0.35, type: "sine", g: 0.07 }); },
    // short pop — item added to cart / scan hit
    add: (c) => { tone(c, { f: 880, d: 0.06, type: "square", g: 0.035 }); tone(c, { f: 1320, t0: 0.045, d: 0.09, type: "sine", g: 0.09 }); },
    // soft double low tone — error (not a buzzer)
    error: (c) => { tone(c, { f: 260, d: 0.14, type: "triangle", g: 0.11, slide: 220 }); tone(c, { f: 230, t0: 0.17, d: 0.2, type: "triangle", g: 0.11, slide: 180 }); },
    // rising sweep — login
    login: (c) => { tone(c, { f: N.G4, d: 0.45, type: "sine", g: 0.12, slide: N.G5 }); tone(c, { f: N.D5, t0: 0.22, d: 0.4, type: "triangle", g: 0.08, slide: N.D5 * 2 }); },
    // gentle two-note chime — critical alert
    alert: (c) => { tone(c, { f: N.B5, d: 0.18, type: "sine", g: 0.12 }); tone(c, { f: N.E5 * 2, t0: 0.16, d: 0.3, type: "sine", g: 0.1 }); },
    // single soft note — ordinary notification / bottom toast
    note: (c) => { tone(c, { f: N.A5, d: 0.16, type: "sine", g: 0.09 }); tone(c, { f: N.A5 * 1.5, t0: 0.05, d: 0.2, type: "sine", g: 0.04 }); },
    // descending minor — invoice voided
    void: (c) => { [N.E5, N.C5, N.A4].forEach((f, i) => tone(c, { f, t0: i * 0.12, d: 0.28, type: "triangle", g: 0.11 })); },
    // hold: two quick descending ticks; resume: reversed
    hold: (c) => { tone(c, { f: N.G5, d: 0.07, type: "sine", g: 0.09 }); tone(c, { f: N.D5, t0: 0.08, d: 0.12, type: "sine", g: 0.09 }); },
    resume: (c) => { tone(c, { f: N.D5, d: 0.07, type: "sine", g: 0.09 }); tone(c, { f: N.G5, t0: 0.08, d: 0.14, type: "sine", g: 0.09 }); },
    // tiny tick — wizard step
    step: (c) => tone(c, { f: 1500, d: 0.05, type: "sine", g: 0.06 }),
    click: (c) => tone(c, { f: 1000, d: 0.03, type: "square", g: 0.025 }),
    // ready — loading finished (short login loading)
    ready: (c) => { tone(c, { f: N.E5, d: 0.12, type: "triangle", g: 0.12 }); tone(c, { f: N.B5, t0: 0.12, d: 0.25, type: "triangle", g: 0.12 }); },
    // install complete — longer, still soft
    install: (c) => { [N.G4, N.C5, N.E5, N.G5, N.C6].forEach((f, i) => tone(c, { f, t0: i * 0.12, d: 0.3, type: "triangle", g: 0.14 })); [N.C5, N.E5, N.G5].forEach((f) => tone(c, { f, t0: 0.7, d: 0.9, type: "sine", g: 0.06 })); },
    // exit — calm descending farewell
    exit: (c) => { [N.C6, N.G5, N.E5, N.C5].forEach((f, i) => pluck(c, f, i * 0.16, 0.5, 0.11)); tone(c, { f: N.C4, t0: 0.5, d: 1.1, type: "sine", g: 0.07, a: 0.04 }); },
  };
  function play(name) {
    if (!enabled()) return false;
    const c = ac(); if (!c) return false;
    if (c.state === "suspended") c.resume().catch(() => {});
    master.gain.value = vol();
    try { (P[name] || P.click)(c); return true; } catch (_) { return false; }
  }
  window.Sfx = { play, enabled, setEnabled: (v) => localStorage.setItem("sm.sfx", v ? "1" : "0"), volume: vol, setVolume: (v) => { localStorage.setItem("sm.sfx.vol", String(v)); if (master) master.gain.value = vol(); }, names: Object.keys(P) };
})();

File: mobile-android/app/src/main/java/ir/khajavy/supermarket/Sfx.java
package ir.khajavy.supermarket;

import android.media.AudioAttributes;
import android.media.AudioFormat;
import android.media.AudioTrack;

/**
 * v2.2 — soft sound effects for the phone, mirroring the PC's frontend/sfx.js.
 * Tones are synthesized (sine/triangle with a gentle envelope), so no audio files
 * are bundled and the Gradle-less build stays tiny. Every level is deliberately
 * low: a shop till must not be loud. Prefs "sfx"="0" disables.
 *
 *  success — sale completed / activation     add   — item added to cart / scan hit
 *  error   — soft low double-tone            note  — ordinary notification
 *  alert   — gentle two-note chime           void  — invoice voided (descending)
 *  hold / resume — invoice parked / restored welcome — short instrumental on start
 */
public final class Sfx {
    private Sfx() {}
    private static final int SR = 22050;
    private static final double C4 = 261.63, E4 = 329.63, F4 = 349.23, G4 = 392, A4 = 440, C5 = 523.25, D5 = 587.33, E5 = 659.25, F5 = 698.46, G5 = 783.99, A5 = 880, B5 = 987.77, C6 = 1046.5;

    public static boolean enabled() { return !"0".equals(Prefs.get("sfx", "1")); }
    public static float volume() { try { return Math.max(0f, Math.min(1f, Float.parseFloat(Prefs.get("sfx_vol", "0.35")))); } catch (Exception e) { return 0.35f; } }

    public static void play(String name) {
        if (!enabled()) return;
        Api.bg(() -> { try { render(name); } catch (Throwable ignore) {} });
    }

    /** one note: freq, start (s), duration (s), gain, 0=sine 1=triangle, optional slide target. */
    private static final class N { final double f, t0, d, g, slide; final int type; N(double f, double t0, double d, double g, int type, double slide) { this.f = f; this.t0 = t0; this.d = d; this.g = g; this.type = type; this.slide = slide; } }
    private static N s(double f, double t0, double d, double g) { return new N(f, t0, d, g, 0, 0); }
    private static N t(double f, double t0, double d, double g) { return new N(f, t0, d, g, 1, 0); }
    private static N sl(double f, double t0, double d, double g, double to) { return new N(f, t0, d, g, 1, to); }

    static N[] pattern(String name) {
        switch (name) {
            case "success": return new N[]{t(C5, 0, .22, .16), t(E5, .09, .22, .16), t(G5, .18, .22, .16), t(C6, .27, .22, .16), s(G5 * 2, .36, .35, .07)};
            case "add": return new N[]{t(880, 0, .06, .05), s(1320, .045, .09, .09)};
            case "error": return new N[]{sl(260, 0, .14, .11, 220), sl(230, .17, .2, .11, 180)};
            case "alert": return new N[]{s(B5, 0, .18, .12), s(E5 * 2, .16, .3, .1)};
            case "void": return new N[]{t(G5, 0, .12, .12), t(E5, .12, .12, .12), t(C5, .24, .3, .12), s(C4, .24, .4, .07)};
            case "hold": return new N[]{s(A5, 0, .09, .09), s(E5, .09, .16, .09)};
            case "resume": return new N[]{s(E5, 0, .09, .09), s(A5, .09, .18, .09)};
            case "welcome": return new N[]{t(C5, 0, .5, .13), t(E5, .18, .5, .13), t(G5, .36, .5, .13), t(C6, .54, .55, .13), t(B5, .9, .5, .12), t(G5, 1.08, .5, .12), t(A5, 1.35, .5, .12), t(C6, 1.53, .7, .12), s(C4, 0, .9, .08), s(G4, .9, .9, .08), s(A4, 1.35, .9, .08), s(F4, 1.98, .9, .08), t(C6, 2.3, 1.0, .11), s(E5, 2.3, 1.0, .05)};
            case "note": default: return new N[]{s(A5, 0, .16, .09), s(A5 * 1.5, .05, .2, .04)};
        }
    }

    private static void render(String name) {
        N[] ns = pattern(name); double end = 0; for (N n : ns) end = Math.max(end, n.t0 + n.d);
        int len = (int) ((end + 0.05) * SR); float[] buf = new float[len]; float vol = volume();
        for (N n : ns) {
            int a = (int) (n.t0 * SR), b = Math.min(len, (int) ((n.t0 + n.d) * SR)); double ph = 0;
            for (int i = a; i < b; i++) {
                double p = (i - a) / (double) (b - a); double f = n.slide > 0 ? n.f * Math.pow(n.slide / n.f, p) : n.f;
                ph += 2 * Math.PI * f / SR; double x = n.type == 1 ? (2 / Math.PI) * Math.asin(Math.sin(ph)) : Math.sin(ph);
                double env = p < 0.06 ? p / 0.06 : Math.pow(1 - (p - 0.06) / 0.94, 2.2);
                buf[i] += (float) (x * env * n.g);
            }
        }
        // soft clip + one-pole low-pass so nothing sounds harsh
        float prev = 0; for (int i = 0; i < len; i++) { float v = buf[i] * vol * 1.6f; v = (float) Math.tanh(v); prev = prev + 0.35f * (v - prev); buf[i] = prev; }
        AudioTrack tr = new AudioTrack.Builder()
                .setAudioAttributes(new AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_NOTIFICATION_EVENT).setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION).build())
                .setAudioFormat(new AudioFormat.Builder().setEncoding(AudioFormat.ENCODING_PCM_FLOAT).setSampleRate(SR).setChannelMask(AudioFormat.CHANNEL_OUT_MONO).build())
                .setBufferSizeInBytes(len * 4).setTransferMode(AudioTrack.MODE_STATIC).build();
        tr.write(buf, 0, len, AudioTrack.WRITE_BLOCKING); tr.play();
        try { Thread.sleep((long) ((end + 0.15) * 1000)); } catch (InterruptedException ignore) {}
        tr.release();
    }
}
